Update from GeoServer translate work

jodygarnett opened the pull request:
Peter has been working on the translate script filling in gaps for docs.geoserver.org - which is much harder than other projects attempted.

* [x] UTF-8 encoding fixes
* [x]  RST admonition fix

petersmythe committed:
Fix UTF-8 encoding issues for international RST files

- Add encoding='utf-8' to all file read operations in translate.py
- Add error handling with errors='replace' for graceful UTF-8 fallbacks
- Fix UnicodeDecodeError on Windows systems with international content
- Addresses CP1252 vs UTF-8 encoding conflicts in RST scanning

petersmythe committed:
Fix UTF-8 encoding issues for international RST files

- Add encoding='utf-8' to all file read operations in translate.py
- Add error handling with errors='replace' for graceful UTF-8 fallbacks
- Fix UnicodeDecodeError on Windows systems with international content
- Addresses CP1252 vs UTF-8 encoding conflicts in RST scanning

petersmythe committed:
Update cli.py

Add missing import errno  
https://github.com/jodygarnett/translate/pull/11

petersmythe committed:
Attempt to run on Windows

petersmythe committed:
Merge branch 'main' of https://github.com/petersmythe/translate

petersmythe committed:
Add --folder arguments to init

petersmythe committed:
Replace HTML line breaks with Markdown line breaks in postprocess_markdown

petersmythe committed:
Replace HTML line breaks with Markdown line breaks in postprocess_markdown

petersmythe committed:
Revert "Add --folder arguments to init"

This reverts commit 87323b735dc96a9f7d301e968cfb41365d34b889.

petersmythe committed:
Update download references handling in scan_download function

petersmythe committed:
Bump version from 0.4.0 to 0.5.0

petersmythe committed:
Improve parameter naming in scan_toctree function

petersmythe committed:
Merge branch 'main' of https://github.com/petersmythe/translate

petersmythe committed:
Update toctree processing for raw markdown output

petersmythe committed:
Enhance error logging in fenced div processing

petersmythe committed:
Enhance error logging in fenced div processing

petersmythe committed:
Enhance fenced div processing with custom title support

petersmythe committed:
Enhance fenced div processing with custom title support

petersmythe committed:
Enhance fenced div processing with stricter type matching

petersmythe committed:
Enhance fenced div processing with stricter type matching

petersmythe committed:
Enhance nested bullet point handling in RST processing

petersmythe committed:
Refactor status variable to state in list table processing

petersmythe committed:
Enhance relative link generation in RST to Markdown conversion

petersmythe committed:
Enhance relative link generation in RST to Markdown conversion

petersmythe committed:
Enhance relative link generation for downloadable files

petersmythe committed:
Improve logging for broken document links

petersmythe committed:
Add mailto protocol for email links in RST processing

petersmythe committed:
Improve content indentation handling for nested directives

petersmythe committed:
Improve content indentation handling for nested directives

petersmythe committed:
Restore RST to HTML conversion for grid cards

petersmythe committed:
Restore RST to HTML conversion for grid cards

petersmythe committed:
Fix HTML tag indentation in RST processing

petersmythe committed:
Enhance RST post-processing to handle Pandoc raw HTML blocks

petersmythe committed:
Enhance RST post-processing to handle Pandoc raw HTML blocks

petersmythe committed:
Bump version to 0.6.0

petersmythe committed:
Fix for windows paths

petersmythe committed:
v0.6.1 increased logging

jodygarnett commented on mkdocs_translate/cli.py:
```diff
@@ -161,7 +162,7 @@ def scan_download( collected: list[str] ):
 
             if download_folder in downloads:
-                downloads[download_folder].extend(download_references)
+                downloads[download_folder].update(download_references)
```
What is happening here? we should name these methods a little better I expect.

petersmythe committed:
Add implementation plan for nested table de-indentation

petersmythe committed:
Add nested table de-indentation preprocessing

- Detect indented list-table directives (indentation > 0)
- Remove indentation to enable proper Pandoc conversion
- Add HTML comments documenting indentation removal
- Fail migration if de-indentation impossible (validation)
- Process all files during migration, not just those in reports

Features:
- detect_nested_tables() finds all indented tables with line numbers
- deindent_nested_table() removes indentation and adds documentation
- Integration in preprocess_rst() before other block directives
- Proper detection avoiding false positives in code examples
- Handles deep nesting (9+ spaces) correctly

Testing:
- test_nested_table_simple.rst: single 3-space indented table
- test_nested_table_multiple.rst: 3 tables at various indent levels
- test_nested_table_deep.rst: deep nesting in admonition
- test_nested_table_mixed.rst: mixed indented/non-indented tables
- test_nested_table_code_example.rst: code blocks not affected
- test_non_indented_table.rst: non-indented tables unchanged

All test cases pass with correct table detection and indentation removal.

petersmythe committed:
Update implementation plan with Phase 5-6 completion

Phases complete:
- Phase 1: Clean branch setup
- Phase 2: Nested table detection implementation
- Phase 3: De-indentation with comments
- Phase 4: Integration into preprocessing pipeline
- Phase 5: Comprehensive test suite (all passing)
- Phase 6: Real-world testing on GeoServer docs

Bugs fixed:
- Code block detection false positives
- Directive line handling in nested contexts

Testing status:
- 6 test files created and passing
- Real GeoServer documentation tested successfully
- Nested tables correctly detected and processed

petersmythe committed:
Complete Phase 7: Documentation and unit tests

Phase 7 deliverables:
- Created NESTED_TABLE_FIX.md (3600+ word technical documentation)
- Updated CHANGES with v0.6.2 entry
- Bumped version to 0.6.2 in __init__.py
- Created tests/test_nested_tables.py with 17 comprehensive tests
- All unit tests passing (100%)

Test coverage:
- 9 tests for detect_nested_tables() function
- 5 tests for deindent_nested_table() function
- 2 integration tests
- 1 edge case test

Documentation includes:
- Problem analysis and root cause
- Solution architecture
- Implementation details
- Testing results
- Troubleshooting guide
- Future improvement suggestions

Ready for Phase 8: Final merge to main branch

petersmythe committed:
Define Phase 8: Final merge strategy

Documented steps for Phase 8:
- Verify branch status
- Clean up temporary files
- Create PR and merge to geoserver-minimal
- Tag version 0.6.2
- Document release notes

Branch readiness: All code complete and tested
Commits on branch: 4 commits from f4e1c22 baseline
Ready for production merge

petersmythe committed:
Fix editable install package discovery

petersmythe committed:
Merge nested-table-deindent into main

petersmythe committed:
Resolve merge conflicts in translate.py

petersmythe committed:
Remove unicode characters from translate.py

petersmythe committed:
Bump version to 0.6.3 and enhance logging in translate.py

petersmythe committed:
Remove redundant debug log for nested table preprocessing in preprocess_rst function

petersmythe committed:
Bump version to 0.6.4 in __init__.py

jodygarnett commented:
@peter-afrigis can we start to merge this bit by bit, rather than have such a long standing thing?

petersmythe commented:
> @peter-afrigis can we start to merge this bit by bit, rather than have such a long standing thing?

48 hours later, I am still fighting nested tables.

petersmythe committed:
v0.6.5: Add comprehensive print and logging statements for execution tracing

jodygarnett commented:
When we get into things like nested tables, it starts to become worthwhile to edit the RST origional.

How many nested tables are there?

And if you are fighting hard it is "faster" to steal a page example over to this project for rapid troubleshooting.

petersmythe committed:
v0.6.7: Add [ENTRY] logging to migrate(), check_folders(), init_anchors(), collect_paths(), collect_path()

petersmythe committed:
v0.6.8: Add module-level [MODULE-INIT], callback logging to trace CLI invocation

petersmythe committed:
v0.6.9: Add STDERR logging to bypass stdout redirection, trace callbacks and main()

petersmythe committed:
v0.6.10: Add [BEFORE-APP-CALL] and [AFTER-APP-CALL] logging to trace typer invocation

petersmythe committed:
v0.6.11: CRITICAL FIX - Add if __name__=='__main__': app() to cli.py to enable migrate command execution

petersmythe committed:
v0.6.12: Remove top-level debugging ([MODULE-INIT], [ENTRY], [DEBUG], [STDERR-*]), keep [v0.6.4] preprocessing markers

petersmythe committed:
v0.6.13: Fix list-table content indentation in _safe_list_table()

- Fixed critical indentation bug in _safe_list_table() function (line 1838)
- Block content now properly indented under directive (3 spaces minimum)
- Removed double {indent} from option formatting
- Ensures pandoc correctly recognizes list-table directives as table structures
- All list-table blocks now render as proper tables, not empty patterns

This fix complements v0.6.11 nested-table fix:
- v0.6.11: De-indents nested list-tables (indent > 0)
- v0.6.13: Indents top-level list-table content under directive

petersmythe committed:
Bump version to 0.6.14 and fix nested table detection

petersmythe committed:
Bump version to 0.6.15 and remove debug print statements from convert_rst() and preprocess_rst()

petersmythe committed:
Bump version to 0.6.16 and remove unnecessary else clause in preprocess_rst()

petersmythe committed:
Bump version to 0.6.17 and add spaces after images

petersmythe committed:
Fix: force image caption newline (two trailing spaces); fallback import for Annotated; add typing-extensions dependency and tests

jodygarnett commented:
@peter-afrigis for https://geoserver.org/geoserver/en/user/extensions/authkey/#installation page the variables are not being filled in like `{{ release }}`.

I think that is done using macros (which makes pages slower then sphinx) so I turned it off by default. To enable it for a page you have to add a snippet to the top of the page.


Generating locally I see the page does contain:
```markdown
---
render_macros: true
---
```

So I expect the substitutions variables changed in sphinx over the time period. I was working to make `main` generate examples that could be used (rather than always recommending SNAPSHOTs).

main `config.py` has:
```
version = pomtree.getroot().find("{http://maven.apache.org/POM/4.0.0}version").text
snapshot = version.find('SNAPSHOT') != -1
if snapshot:
    print("Building snapshot docs version", version)
else:
    print("Building release docs version", version)
    

# The relacement |release| most recent release version, including alpha/beta/rc tags.
# This should be updated after each release
release = '2.28.0'
if not snapshot:
  release = version
...
branch = 'main'
series = '3.x'
...
rst_epilog = "\n" \
 ".. |branch| replace:: "+branch+"\n" \
 ".. |series| replace:: "+series
```

So the valid substations should be:
* branch = main
* series = 3.x
* version = 3.0-SNAPSHOT
* release = 2.28.0 

And using that we should be able to build up the download macros, reading the logic config also has:

* community = 3.0-SNAPSHOT

config.py uses these values for external link generation:

```
'download_extension': (download_extension,'geoserver-'+release+'-%s-plugin.zip')
'nightly_extension': (nightly_extension,'geoserver-'+community+'-%s-plugin.zip'),
```

jodygarnett commented:
There is some logic here for config handling external links:
- https://github.com/petersmythe/translate/blob/main/mkdocs_translate/translate.py#L1212 

And the example `translate.yml` has several:

```
extlinks:
  github: https://github.com/jodygarnett/translate/blob/main/%s
  tag: https://github.com/jodygarnett/translate/releases/tag/{{ release }}|Release {{ release }}
  squidfunk: https://squidfunk.github.io/mkdocs-material/%s
```

But I do not see a translate.yml for external links in the geoserver user docs.

jodygarnett commented:
I tried creating a local translate.yml with extlinks and running locally but did not get anywhere.

Let's catch up, and if needed remove this stuff.
